signer/src/dkg/verification.rs

```diff
@@ -0,0 +1,804 @@
+    SigningFailure(Box<SignError>),
+
+    /// The FROST coordinator returned an unexpected result.
+    #[error("unexpected WSTS result: {0:?}")]
+    UnexpectedWstsResult(Box<OperationResult>),
+
+    /// An error occurred in the WSTS coordinator.
+    #[error("WSTS coordinator error: {0}")]
+    Coordinator(Box<dyn std::error::Error + 'static + Send + Sync>),
+}
+
+/// Represents the state of a DKG verification.
+#[derive(Debug, Clone)]
+pub enum State {
+    /// The DKG verification has been created but not yet been used to process
+    /// any messages.
+    Idle,
+    /// The DKG verification signing round has begun but has not yet finalized.
+    Signing,
+    /// The DKG verification has completed successfully, and the resulting
+    /// signature is stored.
+    Success(TaprootSignature),
+    /// The DKG verification has failed due an an error or unexpected result
+    /// from the FROST coordinator.
+    Error,
+    /// The DKG verification has expired and can no longer be used.
+    Expired,
+}
+
+impl std::fmt::Display for State {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            State::Idle => write!(f, "idle"),
+            State::Signing => write!(f, "signing"),
+            State::Success(_) => write!(f, "success"),
+            State::Error => write!(f, "error"),
+            State::Expired => write!(f, "expired"),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct QueuedMessage {
+    processed: bool,
+    message: wsts::net::Message,
+}
+
+impl QueuedMessage {
+    fn new(message: wsts::net::Message) -> Self {
+        Self { processed: false, message }
+    }
+
+    fn mark_processed(&mut self) {
+        self.processed = true;
+    }
+}
+
+/// Represents the state of a DKG verification. This implementation is designed
+/// to be able to span across multiple Bitcoin blocks, being limited by
+/// wall-clock time instead of Bitcoin block cadence.
+///
+/// TODO: The signer currently stops messages from being processed if the
+/// received message doesn't match the current chain tip, so that needs to be
+/// changed for cross-block functionality to work.
+#[allow(dead_code)]
+#[derive(Debug)]
+pub struct StateMachine<TError> {
+    /// The aggregate key that is being verified.
+    aggregate_key: secp256k1::XOnlyPublicKey,
+    /// The state machine that is being used to verify the signer.
+    coordinator: FrostCoordinator,
+    /// WSTS messages that have been received from other signers. We keep them
+    /// in a [`HashMap`] where the keys are tuples of message types and sender
+    /// public keys, and the values are queued messages. This allows us to
+    /// manage messages by type and sender, ensuring that we can handle
+    /// out-of-order messages and process them correctly, while also
+    /// constraining each signer to one message per message type.
+    wsts_messages: HashMap<(WstsNetMessageType, PublicKey), QueuedMessage>,
```

Comment:
Given the usage we do below something like:
```suggestion
    wsts_messages: HashMap<WstsNetMessageType, HashMap<PublicKey, QueuedMessage>>,
```
could simplify some operations.

Reply:
I had it like that before and switched it to the current form because it actually simplified things a bit :)